### DISCUSSION TITLE
What is best practice to write to audit log?

### DISCUSSION THREAD

**Discussion by natilivni** [dotnet/dotNext#27] [Q&A] [answered]
Hello, regarding the following from issue #5:

"Use IReplicatedCluster interface that can be injected using DI for writing changes into Write Ahead Log. You need to call WriteAsync method"

I can not find this method on the interface nor can I find an extension method for this. Has this been refactored since this issue? If so, what is the current correct way to initiate a write to the audit log?

**Comment by sakno** [comment, marked as answer]
Hi @natilivni 

In context of custom database please examine [this](https://sakno.github.io/dotNext/features/cluster/raft.html#guide-how-to-implement-database) article first.

Issue #5 is no longer relevant because support of 1.x branch has been dropped since 12/20/2020. The currently supported 2.x branch doesn't have this method because `WriteConcern` was removed as a part of the simplification. `WriteAsync` was introduced to ensure that the written log entry is committed by the majority of nodes. Now it's a responsibility of audit log itself. If you need to add a new log entry then do the following calls.
1. `IPersistentState.AppendAsync` to add log entries in _uncommitted_ state to the log, save the returned entry index
1. Optionally, `IReplicationCluster<IRaftLogEntry>.ForceReplicationAsync` to force a replication of the _uncommited_ log entries to other nodes. Without this method, replication will be done automatically due to heartbeat timeout. But if you don't want to wait for the heartbeat then it's possible to trigger replication immediately.
1. `IPersistentState.WaitForCommitAsync` and pass the index of the _uncommitted_ log entry from the step 1

P.S.: I recommend you to check discussion #25 as well.

**Comment by natilivni** [comment]
Thank you! Very helpful. I actually have some more questions but will start other discussions regarding them so that others can benefit.
